### PR TITLE
Convert shared/constants/metametrics to TS

### LINKED ISSUE #18285: Convert NetworkController dependencies to TypeScript
We want to convert NetworkController to TypeScript in order to be able to compare differences between the version of NetworkController in this repo and the version in the core repo. To do this, however, we need to convert the following files:

- `shared/modules/fetch-with-timeout.js`
- `shared/modules/network.utils.js`
- `shared/constants/metametrics.js`

We will also need to:

- upgrade to the latest version of `@metamask/swappable-obj-proxy` in order to be able to use the types there
- provide minimal types for `eth-query`

### PR DESCRIPTION
**Pull request by mcmire**
## Explanation

We want to convert NetworkController to TypeScript in order to be able to compare differences in the controller between in this repo and the core repo. To do this, however, we need to convert the dependencies of the controller to TypeScript.

As a part of this effort, this commit converts `shared/constants/metametrics` to TypeScript. Note that simple objects have been largely replaced with enums. There are some cases where I even split up some of these objects into multiple enums.

Progresses #18285.

## Manual Testing Steps

This commit should not introduce any functional changes. All tests should pass, and all functionality in the extension should continue to work.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: shared/constants/metametrics.ts
Comment (mcmire) [inline shared/constants/metametrics.ts:539]: Note that previously all of these event-related enums were packaged into one object. Enum values must be simple, so we can't nest these into a larger enum.

File: shared/constants/metametrics.ts
Comment (mcmire) [inline shared/constants/metametrics.ts:627]: Previously this enum was nested in another object (`METAMETRIC_KEY_OPTIONS`). Considering that this object was being used for only one thing, I didn't think it was worth keeping that structure. Also, as above, enum values must be simple so this enum follows the same pattern as the enums above.

File: shared/constants/metametrics.ts
Comment (mcmire) [inline shared/constants/metametrics.ts:440]: Once we upgrade TypeScript we should be able to use `satisfies MetaMetricsPageObject` instead of typing it like this, but this should work for the time being.

File: app/scripts/background.js
Comment (mcmire) [inline app/scripts/background.js:27]: Fixed this typo while I was at it.

File: shared/constants/metametrics.ts
Comment (mcmire) [reply shared/constants/metametrics.ts]: Ah yes! I didn't know this existed. Updated in [150a143](https://github.com/MetaMask/metamask-extension/pull/18353/commits/150a1437113799742b1c4d10605a3dd971007643).

File: shared/constants/metametrics.ts
Comment (mcmire) [reply shared/constants/metametrics.ts]: Good catch! Fixed in [3f7711b](https://github.com/MetaMask/metamask-extension/pull/18353/commits/3f7711bdb60f8f37bb1fc340d4d02f98333edf0f).

File: ui/components/app/detected-token/detected-token.js
Comment (Gudahtt) [inline ui/components/app/detected-token/detected-token.js:98]: Looks like the property was forgotten here:
```suggestion
        location: MetaMetricsEventLocation.TokenDetection,
```

**Comment by Gudahtt** [comment]
Looks great! I can approve after the one problem I found is resolved

**Review by Gudahtt — DISMISSED** [review]
LGTM!

File: ui/store/action-queue/index.ts
Comment (mcmire) [reply ui/store/action-queue/index.ts:195]: I went ahead and fixed this instance since it was used in two places: [83b3f0a](https://github.com/MetaMask/metamask-extension/pull/18353/commits/83b3f0af9bcb621b7921f087352bf9fb34affaf4)

Comment (Gudahtt) [reply ui/store/action-queue/index.ts:195]: Makes sense!

**Comment by mcmire** [comment]
CI looked like it was about to pass before I merged `develop` again, so I don't anticipate any more changes.

**Review by Gudahtt — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]
LGTM, thanks @mcmire

**Pull request merged**
